### ISSUE TITLE
Fixes for Xubuntu 24.04

### ISSUE THREAD

**Issue by capnbfg** [joshbarrass/ut2003-on-linux#2]
Thanks for creating this guide and Lutris script.  I did use your install script, but it broke for me at the padsp step.  Here are all the problems I encountered and the changes I'd suggest to work through it.

BACKGROUND / ISSUES:

This command worked fine for me:
sudo apt-get install libstdc++5:i386

However, this package does not exist for 24.04/Noble:
sudo apt-get install libpulsedsp:i386
_E: Unable to locate package libpulsedsp:i386_

If I try running 'padsp', I get the following suggestion:
_Command 'padsp' not found, but can be installed with:
sudo apt install pulseaudio-utils_

As you noted in the instructions, we actually need the i386 version.  This works for me:
sudo apt install pulseaudio-utils:i386

However, it does not create a padsp script in any executable search path.  As a result, your Lutris script dies at the following step:
_command: cp "$(which padsp)" "$GAMEDIR/padsp32"_

Even after installing the above packages, /dev/dsp does not exist, so I still get the following error:
_open /dev/[sound/]dsp: No such file or directory_

I also had a problem with an invalid cdkey file being created.  My display driver echos some text at runtime:
_radeonsi: Forcing anisotropy filter to 16x_
This text was somehow captured into System/cdkey on a line preceding the actual key, which subsequently caused a game crash.  Fortunately there was a message in the Lutris log file stating that the cdkey was invalid, so this was easy to catch.

FIXES / WORKAROUNDS:

Although the install script died, it was late enough in the flow that the installation was salvageable.
I manually created $GAMEDIR/launch-ut2003.sh and then added to Lutris as a locally installed game.

I ran a find operation for padsp from /, which located a usable script here:
/snap/gnome-3-28-1804/198/usr/bin/padsp

 I copied the above padsp to $GAMEDIR/padsp32 and modified the LD_PRELOAD as instructed.  However, I had to change the .so name as follows:
`if [ x"$LD_PRELOAD" = x ] ; then
    LD_PRELOAD="/usr/lib/i386-linux-gnu/pulseaudio/libpulsecommon-16.1.so"
else
    LD_PRELOAD="$LD_PRELOAD /usr/lib/i386-linux-gnu/pulseaudio/libpulsecommon-16.1.so"
fi`

To create /dev/dsp, I needed to install another package:
sudo apt install osspd

The cdkey issue was easy to fix.  I just had to open the file in a text editor and delete the extra lines.

CONCLUSION:

After the above steps, I now have the game running well and with working audio.

I still have a few graphical things to tune up before I'd call it perfect, but for this issue I'm just focused on the showstoppers. I will share any additional optimizations as a separate issue.

**Comment by narekavagyan** [comment]
Any sound solution for Ubuntu 24.04.3 LTS? Also game screen is smaller than my screen, couldn't fix that too.

E: Unable to locate package pulseaudio-util:i386

Thanks!

**Comment by capnbfg** [comment]
You're right, I had a typo there (fixed now).  It should have said:
sudo apt install pulseaudio-util**s**:i386

**Comment by capnbfg** [comment]
I handled graphical tweaks in two places.

The first place is inside my launch script.  Here are the complete contents of that script, which you can use as a starting point.

If you're choosing to run a lower resolution or a 4:3 resolution on a modern display, then you'll want to adjust how that image is scaled.  On Xubuntu (xfce desktop), the utility for this is "xrandr", but since you're on Ubuntu (Gnome desktop) you'll need to figure out the equivalent utility and options.

```
#!/bin/bash

ABSDIR=$(readlink -f $0)
GAMEDIR=$(dirname $ABSDIR)
echo $GAMEDIR


DRIVER_IS_MESA=$(glxinfo -B | grep -i 'renderer info' | grep -ic mesa)
OS_IS_UBUNTU=$(grep PRETTY_NAME /etc/os-release | grep -ic ubuntu)
DE_IS_XFCE=$(ls /usr/bin/*session | grep -ic xfce)


if [[ ! -e "$GAMEDIR/System/cdkey" ]]; then
    zenity \
        --icon-name="lutris" \
        --width="450" \
        --height="250" \
        --title="Unreal Tournament 2003" \
        --entry \
        --text "Please enter your UT2003 CD key\nYou can change this at a later date by editing $GAMEDIR/System/cdkey" \
        > "$GAMEDIR/System/cdkey"
fi


cd "$GAMEDIR/System"


if [[ $DRIVER_IS_MESA ]]; then
    # force 16x anisotropic filtering
    export R600_TEX_ANISO=16
    export RADV_TEX_ANISO=16
    export AMD_TEX_ANISO=16
    # force anti-aliasing (range 2-32)
    export pp_jimenezmlaa_color=16
fi


# force centered rendering to preserve size and aspect
if [[ $OS_IS_UBUNTU && $DE_IS_XFCE ]]; then
    for output in $(xrandr --prop | grep -E -o -i "^[A-Z\-]+-[0-9]+"); do
      xrandr --output "$output" --set "scaling mode" "Center"
      xrandr --output "$output" --set "TearFree"     "on"
    done
fi


# run the game with pulse audio 32-bit legacy sound wrapper
../padsp32 ./ut2003-bin
RETCODE=$?


# revert system to default scaling mode (none)
if [[ $OS_IS_UBUNTU && $DE_IS_XFCE ]]; then
    for output in $(xrandr --prop | grep -E -o -i "^[A-Z\-]+-[0-9]+"); do
      xrandr --output "$output" --set "scaling mode" "None"
      xrandr --output "$output" --set "TearFree"     "off"
    done
fi


# bubble up the return code from the game
if [[ $RETCODE -ne 0 ]]; then
    echo "Game process exited with retCode $RETCODE"
    exit $RETCODE
fi
```


The second place to customize graphical settings is inside the game's ini file. Look for ~/.ut2003/System/UT2003.ini, which is created after first launch of the game.  Initially I set 1280x1024 resolution and used xrandr to force centered scaling, but later I found that my monitor's native 1920x1080 resolution worked just fine.

Below are the settings that I ended up with.  I do not recommend copying these verbatim because some of these settings might need to be different for your system, but you can use this as a reference to see what worked for me.

```
[SDLDrv.SDLClient]
WindowedViewportX=1920
WindowedViewportY=1080
FullscreenViewportX=1920
FullscreenViewportY=1080
MenuViewportX=1024
MenuViewportY=768
Brightness=0.800000
Contrast=0.700000
Gamma=0.800000
UseJoystick=False
JoystickNumber=0
IgnoreHat=False
JoystickHatNumber=0
CaptureMouse=True
StartupFullscreen=True
ScreenFlashes=True
NoLighting=False
MinDesiredFrameRate=60.000000
Decals=True
Coronas=True
DecoLayers=True
Projectors=True
NoDynamicLights=False
ReportDynamicUploads=False
TextureDetailInterface=Normal
TextureDetailTerrain=UltraHigh
TextureDetailWeaponSkin=UltraHigh
TextureDetailPlayerSkin=UltraHigh
TextureDetailWorld=UltraHigh
TextureDetailRenderMap=UltraHigh
TextureDetailLightmap=UltraHigh
TextureMaxLOD=0
TextureMinLOD=0
NoFractalAnim=False
```

```
[OpenGLDrv.OpenGLRenderDevice]
MaxTextureUnits=8
VARSize=32
ReduceMouseLag=False
AppleVA=0
MultisampleBuffers=0
MultisampleSamples=0
MultisampleHint=2
UseRenderTargets=False
ForceCompression=False
TerrainLOD=0
SkyboxHack=False
LowQualityTerrain=False
DetailTextures=True
DetailTexMipBias=0.000000
HighDetailActors=True
SuperHighDetailActors=True
UsePrecaching=True
UseTrilinear=True
UseCompressedLightmaps=True
UseStencil=False
UseCubemaps=False
LevelOfAnisotropy=0.000000
DesiredRefreshRate=144
UseVSync=False
UseTripleBuffering=False
Use16bitTextures=False
Use16bit=False
```

I remember the ReduceMouseLag=False setting being a eureka moment, but now after all these months I cannot remember what issue that setting fixed 😆

**Comment by narekavagyan** [comment]
Thank you so much for your help. Can you also edit exact commands after installing pulseaudio-utils:i386?

**Comment by capnbfg** [comment]
Essentially all that I did was follow the existing guide [here](https://github.com/joshbarrass/ut2003-on-linux) and improvised on the steps where I encountered issues.  I can't guarantee an exact step-by-step process because too much time has passed and my system isn't identical to yours, but here's an approximate order of operations.

Ensure that you have installed all the requisite packages:
`sudo apt-get install libstdc++5:i386 pulseaudio-utils:i386 osspd`

You need a padsp script to use when starting the game.  Per the original guide, that script should have been created in your executable $PATH by installing pulseaudio-utils, but for me it was not.  You can check by doing
`which padsp`

Your system may be slightly different, but it should be located approximately here:
`ls -al /snap/gnome-3-28-1804/198/usr/bin/padsp`

Once you find the script, copy it to your game directory, like this:
`cp /snap/gnome-3-28-1804/198/usr/bin/padsp $GAMEDIR/padsp32`

Then use your text editor of choice open the padsp32 script and correct the LD_PRELOAD paths like this:
`gedit $GAMEDIR/padsp32`
```
if [ x"$LD_PRELOAD" = x ] ; then
   #LD_PRELOAD="/usr/lib/x86_64-linux-gnu/pulseaudio/libpulsedsp.so"
   LD_PRELOAD="/usr/lib/i386-linux-gnu/pulseaudio/libpulsecommon-16.1.so"
else
   #LD_PRELOAD="$LD_PRELOAD /usr/lib/x86_64-linux-gnu/pulseaudio/libpulsedsp.so"
   LD_PRELOAD="$LD_PRELOAD /usr/lib/i386-linux-gnu/pulseaudio/libpulsecommon-16.1.so"
fi
```

I also recommend replacing the exec line inside padsp32 with a different syntax that will bubble up any error codes from the game.
```
# replace this:
#exec "$@"

# with this:
"$@"
RETCODE=$?
exit $RETCODE
```

Now that your padsp script is created, you need a launch script for Lutris to use.
```
touch $GAMEDIR/launch-ut2003.sh
chmod +x $GAMEDIR/launch-ut2003.sh
```

Then edit the script with your text editor of choice:
`gedit $GAMEDIR/launch-ut2003.sh`

Paste in the contents of the launch script that I shared earlier.  You could remove the xfce-specific parts of my script, or you could replace them with equivalent code that works on Gnome.  I personally don't use Gnome, so I can't help you there.

Once the script is done, you need to tell Lutris to use this script as the executable for the game.  In Lutris, right click on the game, go to Configure, go to Game Options, and in the Executable box enter the path to your launch-ut2003.sh script.  For me, that looks like this:
`/data/Games/unreal-tournament-2003/launch-ut2003.sh`

Now try running the game.  On the first run, I got an error about the cdkey being invalid, but it was because incorrect text had found its way into the auto-generated key file.  If you see any such error, open $GAMEDIR/System/cdkey in your text editor of choice and correct it such that the first line of the file contains nothing but your cdkey text.

And...that's it.

**Comment by narekavagyan** [comment]
nothing after which padsp. Thanks for your effort though

**Comment by narekavagyan** [comment]
Also regarding video fix, my game launcher is u2003-bin file in System folder. Not sure how to edit binary file. Or do you mean the linux_installer.sh file?

Thanks!
